f3.go

```diff
@@ -145,412 +78,197 @@ func New(ctx context.Context, manifest manifest.ManifestProvider, ds datastore.D
-			continue
-		}
-		select {
-		case queue <- gmsg:
-		case <-ctx.Done():
-			break loop
+	// Try to start immediately if we have a manifest available and don't have to wait to
+	// bootstrap. That way, we'll be fully started when we return from Start.
+	if initialDelay == 0 {
+		if err := m.reconfigure(startCtx, pendingManifest); err != nil {
+			return xerrors.Errorf("failed to start GPBFT: %w", err)
 		}
+		pendingManifest = nil
 	}
-}
 
-// Callback to be triggered when there is a dynamic manifest change
-// If the manifest triggers a rebootstrap it starts a new runner with the new configuration.
-// If there is no rebootstrap it only starts a new pubsub topic with a new network name that
-// depends on the manifest version so there is no overlap between different configuration instances
-func ManifestChangeCallback(m *F3) manifest.OnManifestChange {
-	// We can only accommodate 1 manifest update at a time, thus
-	// the buffer size.
-	manifestUpdate := make(chan uint64, 1)
-	m.client.manifestUpdate = manifestUpdate
-	return func(ctx context.Context, prevManifest manifest.Manifest, errCh chan error) {
-		// Tear down pubsub.
-		if err := m.teardownPubsub(prevManifest); err != nil {
-			// for now we just log the error and continue.
-			// This is not critical, but alternative approaches welcome.
-			m.log.Errorf("error stopping gpbft runner: %+v", err)
+	m.errgrp.Go(func() error {
+		manifestChangeTimer := time.NewTimer(initialDelay)
+		if pendingManifest == nil && !manifestChangeTimer.Stop() {
+			<-manifestChangeTimer.C
 		}
-		// empty message queue from outstanding messages
-		m.emptyMessageQueue()
 
-		if m.Manifest().Pause {
-			m.pauseCallback()
-		} else {
+		defer manifestChangeTimer.Stop()
 
-			if m.Manifest().ReBootstrap {
-				m.withRebootstrapCallback(ctx, errCh)
-			} else {
-				m.withoutRebootstrapCallback(ctx, manifestUpdate, errCh)
+		for m.runningCtx.Err() == nil {
+			select {
+			case update := <-m.manifestProvider.ManifestUpdates():
+				if pendingManifest != nil && !manifestChangeTimer.Stop() {
+					<-manifestChangeTimer.C
+				}
+
+				pendingManifest = update
+				if delay, err := m.computeBootstrapDelay(update); err != nil {
+					return err
+				} else if delay > 0 {
+					manifestChangeTimer.Reset(delay)
+					continue
+				}
+			case <-manifestChangeTimer.C:
+			case <-m.runningCtx.Done():
+				return nil
+			}
+			if err := m.reconfigure(m.runningCtx, pendingManifest); err != nil {
+				return xerrors.Errorf("failed to reconfigure F3: %w", err)
 			}
+			pendingManifest = nil
 		}
+		return nil
+	})
 
-	}
+	return nil
 }
 
-func (m *F3) withRebootstrapCallback(ctx context.Context, errCh chan error) {
-	m.mLk.Lock()
-	m.log.Infof("triggering manifest (seq=%d) change with rebootstrap", m.Manifest().Sequence)
-
-	// if runner still up
-	if m.runner != nil {
-		m.runner.Stop()
-	}
-
-	// clear the certstore
-	if err := m.certStore.DeleteAll(ctx); err != nil {
```

Comment:
We now use a new namespace each time. I figure:

1. We'll clear the old datastore when we turn everything on for real.
2. We _should_ use a separate top-level `/f3` datastore to avoid polluting the other lotus datastores.